### PR TITLE
Issue1248: Persistent Email Reminders

### PR DESCRIPTION
**Pull request by anushka-singh**
### Feature or Bugfix
- Feature

### Detail
- When a share request is initiated and remains pending for an extended period, dataset producers will receive automated email reminders at predefined intervals. These reminders will prompt producers to either approve or extend the share request, thereby preventing delays in accessing datasets.

Attaching screenshots for emails:

Screenshot 2024-06-20 at 5 34 31 PM

Screenshot 2024-06-20 at 5 34 52 PM

- Email will be sent every Monday at 9am UTC. Schedule can be changed in cron expression in container.py

### Relates
- https://github.com/data-dot-all/dataall/issues/1248

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: deploy/stacks/container.py [hunk 1]
```diff
@@ -44,11 +47,20 @@ def __init__(
         self._ecr_repository = ecr_repository
         self._vpc = vpc
         self._prod_sizing = prod_sizing
+        self._email_custom_domain = email_custom_domain
+        self._ses_configuration_set = ses_configuration_set
```
Comment (dlpzx) [inline deploy/stacks/container.py:51]: The object variables `self._email_custom_domain` and `self._ses_configuration_set` are not used anywhere, we can remove them

File: deploy/stacks/container.py [hunk 2]
```diff
@@ -44,11 +47,20 @@ def __init__(
...
 
         (self.scheduled_tasks_sg, self.share_manager_sg) = self.create_ecs_security_groups(
             envname, resource_prefix, vpc, vpce_connection, s3_prefix_list, lambdas
         )
         self.ecs_security_groups: [aws_ec2.SecurityGroup] = [self.scheduled_tasks_sg, self.share_manager_sg]
+        self.env_vars = self._create_env('INFO')
```
Comment (dlpzx) [inline deploy/stacks/container.py:57]: Maybe we can move this logic inside the creation of the task in `add_persistent_email_reminders_task` and make it a local variable inside the function

File: backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import logging
+import os
+import sys
+from dataall.modules.dataset_sharing.api.types import ShareObject
```
[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py:4]: `from dataall.modules.dataset_sharing.api.types import ShareObject` is wrongly imported:
1. It should not be a graphql object, we are just doing a query to a table in pending_shares. Instead it should be a db.model
2. The db.model should be imported from `shares_base`

`from dataall.modules.shares_base.db.share_object_models import ShareObject`

File: backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def persistent_email_reminders(engine, envname):
+    """
+    A method used by the scheduled ECS Task to run persistent_email_reminder() process against ALL
+    active share objects within data.all and send emails to all pending shares.
+    """
+    with engine.scoped_session() as session:
+        log.info('Running Persistent Email Reminders Task')
+        pending_shares = SharingService.fetch_pending_shares(engine=engine)
```
[the remark this answers is not in the document]
Comment (dlpzx) [reply backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py:25]: We should also rename it to something that indicates that we are also joining the Notifications, it is not just listing the pending shares

File: backend/dataall/modules/shares_base/handlers/ecs_share_handler.py [hunk 5]
```diff
@@ -30,6 +30,13 @@ def verify_share(engine, task: Task):
     def reapply_share(engine, task: Task):
         return EcsShareHandler._manage_share(engine, task, SharingService.reapply_share, 'reapply_share')
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/handlers/ecs_share_handler.py:32]: This worker is not used right? I think we do not need it because the ECS task for persistent emails is scheduled. The worker decorator registers this handler in the Worker Lambda which would trigger the ECS task on demand

File: backend/dataall/modules/shares_base/services/sharing_service.py [hunk 6]
```diff
@@ -356,6 +362,48 @@ def reapply_share(cls, engine: Engine, share_uri: str) -> bool:
                     if not lock_released:
                         log.error(f'Failed to release lock for dataset {share_data.dataset.datasetUri}.')
 
+    @staticmethod
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/sharing_service.py:365]: Already commented above, `fetch_pending_shares` should be part of a db.repository

File: backend/dataall/modules/shares_base/services/sharing_service.py [hunk 7]
```diff
@@ -356,6 +362,48 @@ def reapply_share(cls, engine: Engine, share_uri: str) -> bool:
...
+                    and_(
+                        ShareObject.shareUri == func.split_part(Notification.target_uri, '|', 1),
+                        ShareObject.datasetUri == func.split_part(Notification.target_uri, '|', 2),
+                    ),
+                )
+                .filter(and_(Notification.type == 'SHARE_OBJECT_SUBMITTED', ShareObject.status == 'Submitted'))
+                .all()
+            )
+            return pending_shares
+
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/sharing_service.py:385]: The SharingService was conceived to manage actual AWS calls to shares (verify, approve...) share_notifications_emails do not really fit into this category. I think we can simplify the implementation by moving the logic of `_get_share_data` directly to the task and call the ShareNotificationService directly.  We would remove all the code in SharingService and have a task looking something like this:

```
def persistent_email_reminders(engine, envname):
    """
    A method used by the scheduled ECS Task to run persistent_email_reminder() process against ALL
    active share objects within data.all and send emails to all pending shares.
    """
    with engine.scoped_session() as session:
        log.info('Running Persistent Email Reminders Task')
        pending_shares = SharingService.fetch_pending_shares(engine=engine)
        log.info(f'Found {len(pending_shares)} pending shares')
        pending_share: ShareObject
        for pending_share in pending_shares:
            log.info(f'Sending Email Reminder for Share: {pending_share.shareUri}')
            SharingService.persistent_email_reminder(uri=pending_share.shareUri, envname=envname)
        log.info('Completed Persistent Email Reminders Task')


```

File: backend/dataall/modules/shares_base/services/sharing_service.py [hunk 8]
```diff
@@ -356,6 +362,48 @@ def reapply_share(cls, engine: Engine, share_uri: str) -> bool:
...
+    @staticmethod
+    def _get_share_data(session, uri):
+        share = ShareObjectRepository.get_share_by_uri(session, uri)
+        dataset = DatasetRepository.get_dataset_by_uri(session, share.datasetUri)
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/sharing_service.py:389]: - DatasetRepository should be DatasetBaseRepository from datasets_base

File: backend/dataall/modules/shares_base/services/sharing_service.py [hunk 9]
```diff
@@ -356,6 +362,48 @@ def reapply_share(cls, engine: Engine, share_uri: str) -> bool:
...
+        share_items_states = ShareObjectRepository.get_share_items_states(session, uri)
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/sharing_service.py:390]: ShareObjectRepository.get_share_items_states does not exist anymore

File: backend/dataall/modules/shares_base/services/sharing_service.py [hunk 10]
```diff
@@ -356,6 +362,48 @@ def reapply_share(cls, engine: Engine, share_uri: str) -> bool:
...
+        """
+        A method used by the scheduled ECS Task to send email notifications to the requestor of the share object
+        """
+        engine = get_engine(envname=envname)
+
+        with engine.scoped_session() as session:
+            share, dataset, states = cls._get_share_data(session, uri)
+            ShareNotificationService(session=session, dataset=dataset, share=share).notify_persistent_email_reminder(
+                email_id=share.owner, engine=engine
+            )
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/sharing_service.py:404]: Let's see why you need to pass the engine...

File: backend/dataall/modules/shares_base/services/share_notification_service.py [hunk 11]
```diff
@@ -56,11 +56,50 @@ def notify_share_object_submission(self, email_id: str):
         self._create_notification_task(subject=subject, msg=email_notification_msg)
         return notifications
 
+    def notify_persistent_email_reminder(self, email_id: str, engine):
+        share_link_text = ''
+        if os.environ.get('frontend_domain_url'):
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/share_notification_service.py:61]: ShareNotificationService is a generic service that can be used in APIs/Worker... We should not assume that `os.environ.get('frontend_domain_url'):` is part of the env_vars. Maybe we should pass it as a function variable

File: backend/dataall/modules/shares_base/services/share_notification_service.py [hunk 12]
```diff
@@ -175,3 +214,40 @@ def _create_notification_task(self, subject, msg):
                     log.info(f'Notification type : {share_notification_config_type} is not active')
         else:
             log.info('Notifications are not active')
+
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/share_notification_service.py:217]: I think we need to review the design a bit because this pattern is new to data.all:
1. an scheduled ECS task gets triggered, reads some metadata from RDS, executes some light business logic and queues Worker tasks
2. the Worker Lambda sends the emails

At first sight it seems like we can simplify the workflow and run everything from ECS

**Review by dlpzx — CHANGES_REQUESTED** [review]
Left some comments and changes needed

File: deploy/stacks/container.py (hunk 1 above)
Comment (anushka-singh) [reply deploy/stacks/container.py:51]: Done!

File: backend/dataall/modules/shares_base/services/sharing_service.py (hunk 9 above)
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/sharing_service.py:390]: Done

File: backend/dataall/modules/shares_base/services/sharing_service.py (hunk 8 above)
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/sharing_service.py:389]: Done

File: backend/dataall/modules/shares_base/services/sharing_service.py (hunk 7 above)
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/sharing_service.py:385]:
```
    @staticmethod
    @Worker.handler(path='ecs.share.persistent_email_reminder')
    def persistent_email_reminder(engine, task: Task):
        return EcsShareHandler._manage_share(
            engine, task, SharingService.persistent_email_reminder, 'persistent_email_reminder'
        )
```
`SharingService.persistent_email_reminder` is used in ecs_share_handler.py like this. I like your suggestion to move it out of sharing service, but how would we define the ecs task in the ecs_share_handler.py then?

File: backend/dataall/modules/shares_base/services/sharing_service.py (hunk 10 above)
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/sharing_service.py:404]: Yes, thats right!

File: backend/dataall/modules/shares_base/services/share_notification_service.py (hunk 11 above)
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:61]: Hmm, we have similar code in `notify_share_object_submission` function. 
```
    def notify_share_object_submission(self, email_id: str):
        share_link_text = ''
        if os.environ.get('frontend_domain_url'):
            share_link_text = f'<br><br> Please visit data.all <a href="{os.environ.get("frontend_domain_url")}/console/shares/{self.share.shareUri}">share link </a> to take action or view more details'
        msg = f'User {email_id} SUBMITTED share request for dataset {self.dataset.label} for principal {self.share.principalId}'
        subject = f'Data.all | Share Request Submitted for {self.dataset.label}'
        email_notification_msg = msg + share_link_text

        notifications = self.register_notifications(
            notification_type=DataSharingNotificationType.SHARE_OBJECT_SUBMITTED.value, msg=msg
        )

        self._create_notification_task(subject=subject, msg=email_notification_msg)
        return notifications
```

Hence, I thought it might be okay to re-use it. LMK if that doesnt work and what might be a better way to address it.

File: backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py:25]: That makes sense! I moved it to shares_base.db.share_object_repositories. 

Did a merge from main.

Renamed function to `fetch_submitted_shares_with_notifications`

File: backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py:4]: Okay, resolved!
I feel like I need a session with you to understand all the new changes to files and how the project is more recently structured. That will help in future development. We can discuss in office hours.

[the remark this answers is not in the document]
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/tasks/persistent_email_reminders_task.py:4]: > This PR is a bit behind from the changes implemented in `main` refactoring shares_base and s3_datasets_shares. I have added a note in #1359 to not introduce more changes in the shares modules until this PR is merged.

Thank you!! It was feeling a bit redundant to keep re-syncing and pulling latest changes and then falling behind again.

File: backend/dataall/modules/shares_base/services/sharing_service.py (hunk 6 above)
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/sharing_service.py:365]: Moved it there.

File: backend/dataall/modules/shares_base/services/share_notification_service.py (hunk 12 above)
Comment (noah-paige) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:217]: I think the only benefit of sending task to queue for worker lambda is if the worker lambda has code bundled in its image that is not there for the ECS container. However this is not the case and both images should have all the backend code at their disposal for whatever data.all task

While separating out all of our backend code into a more microservice aligned design could be useful - it is not in scope for this PR and requires much more design consideration and careful thought

TLDR - I think let's just keep it all in ECS :)

File: deploy/stacks/container.py (hunk 2 above)
Comment (anushka-singh) [reply deploy/stacks/container.py:57]: We could, but environment variables are usually defined in __init__ right? I dont like the idea so much of it being moved inside the add_persistent_email_reminders_task function

File: backend/dataall/modules/shares_base/services/share_notification_service.py (hunk 12 above)
Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:217]: I actually had discussed the design with Noah before implementing it and we considered both options:

Option 1: Create a new ECS task that finds pending shares, creates tasks for them, and queues them up, then uses the existing worker Lambda to send emails.

Option 2: Use the send_email_task directly, as you suggested.

I chose Option 1 because I thought having a queue would be beneficial for managing a high volume of outgoing emails. It helps to queue up tasks and avoid potential clashes.

I am open to hearing what others have to say too.

**Comment by anushka-singh** [comment]
I am not sure why, but github is not letting me push more commits onto this branch for a PR. 
https://github.com/data-dot-all/dataall/compare/main...anushka-singh:aws-dataall:main?expand=1
Its making me create a brand new PR like this: https://github.com/data-dot-all/dataall/compare/main...anushka-singh:aws-dataall:main?expand=1

**Comment by anushka-singh** [comment]
@dlpzx @noah-paige 
Addressed all the comments. Please take a look again.

File: deploy/stacks/container.py (hunk 2 above)
Comment (dlpzx) [reply deploy/stacks/container.py:55]: Convinced :)

File: backend/dataall/modules/shares_base/services/share_notification_service.py (hunk 11 above)
Comment (dlpzx) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:61]: Looking at the rest of the code, if it is not present the message will still be sent (just without the redirect link). So it is not catastrophical that it is not passed. We can leave it as it is for now

File: backend/dataall/modules/shares_base/services/share_notification_service.py (hunk 12 above)
Comment (dlpzx) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:206]: I am going to vote for option 2: If we were implementing this from scratch, I would go for a queue + Lambda to process the emails. The part that I do not like form the current architecture is that we are using an ECS task to do very minor things. The new pattern that I would like to see is: scheduled Lambda ---> SQS queue ---> Worker Lambda. But that is out of scope for this PR. We can approve the current scheduled ECS ---> SQS queue ---> Worker Lambda and think about scheduled ECS tasks that could be handled in Lambda in a separate PR

**Review by dlpzx — APPROVED** [review]

File: backend/dataall/modules/shares_base/services/share_notification_service.py (hunk 12 above)
Comment (noah-paige) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:206]: +1 after some more thought I think we maybe should just keep all compute for email reminders in ECS for now

I think it should only require a small change to L238 of `share_notification_service` to not use `Worker.queue()` and instead directly call 
```
SESEmailNotificationService.send_email_task(
     subject, message, recipient_groups_list, recipient_email_list
 )
```

I think it could be worthwhile to keep Task creation nonetheless for auditability

Comment (anushka-singh) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:207]: @noah-paige Added 
```
SESEmailNotificationService.send_email_task(
     subject, message, recipient_groups_list, recipient_email_list
 )
```
directly to `share_notification_service`

Comment (dlpzx) [reply backend/dataall/modules/shares_base/services/share_notification_service.py:207]: typo: I voted for option 1, but I am not strong opinionated.

File: backend/dataall/modules/shares_base/services/share_notification_service.py [hunk 13]
```diff
@@ -235,7 +236,9 @@ def _create_persistent_reminder_notification_task(self, subject, msg, engine):
                         self.session.add(notification_task)
                         self.session.commit()
 
-                        Worker.queue(engine=engine, task_ids=[notification_task.taskUri])
+                        SESEmailNotificationService.send_email_task(
```
Comment (dlpzx) [inline backend/dataall/modules/shares_base/services/share_notification_service.py:239]: We do not need to send the engine to the function anymore :)

**Review by noah-paige — APPROVED** [review]
